### ISSUE TITLE
By setting seed, tensor.uniform_() still generate different random matrix

### ISSUE THREAD

**Issue by hannshu** [pytorch/pytorch#97129]
### 🐛 Describe the bug

If I define two fc layers, it should generate the same matrix no matter how many times I run. But if I delete the later fc layer, it appears that it cannot generate the same matrix as before.

``` python
import torch
import torch.nn as nn

# set seed
torch.manual_seed(0)
print(torch.initial_seed())

class model(nn.Module):

    def __init__(self) -> None:
        super().__init__()

        self.fc1 = nn.Linear(in_features=3, out_features=2, bias=False)
        self.fc2 = nn.Linear(in_features=3, out_features=3, bias=False)

        self.reset_parameters()


    def reset_parameters(self):
        torch.nn.init.xavier_normal_(self.fc1.weight.data, gain=1)
        torch.nn.init.xavier_normal_(self.fc2.weight.data, gain=1)

m = model().to('cuda')

print(m.fc1.weight.data)
print(m.fc2.weight.data)
print(torch.initial_seed())
```
result:  
``` bash
$$ python3 test_reset.py
0
tensor([[ 0.3817,  0.5129, -0.0285],
        [ 0.5564,  0.6629, -0.0281]], device='cuda:0')
tensor([[-0.4174,  1.6549, -0.3265],
        [ 0.0926, -0.0146,  0.6200],
        [ 1.3065, -0.5297, -0.1300]], device='cuda:0')
0
```
when I delete self.fc2:  
``` python
import torch
import torch.nn as nn

# set seed
torch.manual_seed(0)
print(torch.initial_seed())

class model(nn.Module):

    def __init__(self) -> None:
        super().__init__()

        self.fc1 = nn.Linear(in_features=3, out_features=2, bias=False)
        # self.fc2 = nn.Linear(in_features=3, out_features=3, bias=False)

        self.reset_parameters()


    def reset_parameters(self):
        torch.nn.init.xavier_normal_(self.fc1.weight.data, gain=1)
        # torch.nn.init.xavier_normal_(self.fc2.weight.data, gain=1)

m = model().to('cuda')

print(m.fc1.weight.data)
# print(m.fc2.weight.data)
print(torch.initial_seed())
```
result:  
``` bash
$$ python3 test_reset.py
0
tensor([[ 0.7997, -0.4348,  0.1014],
        [-0.3836, -0.4953,  0.6718]], device='cuda:0')
0
```
You can see that this fc1 layer has the same size and the same seed, but those two layers do not have the same random matrix.  
I do not know whether this is a deliberately added feature or a bug. If this is a bug, I think there may be something wrong with `tensor.uniform_()`.

### Versions

``` bash
$$ python3 collect_env.py 
Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:39:03)  [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-40-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 10.1.243
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti
GPU 2: NVIDIA GeForce RTX 2080 Ti
GPU 3: NVIDIA GeForce RTX 2080 Ti
GPU 4: NVIDIA GeForce RTX 2080 Ti
GPU 5: NVIDIA GeForce RTX 2080 Ti
GPU 6: NVIDIA GeForce RTX 2080 Ti
GPU 7: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 465.27
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          72
On-line CPU(s) list:             0-71
Thread(s) per core:              2
Core(s) per socket:              18
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Gold 6254 CPU @ 3.10GHz
Stepping:                        7
CPU MHz:                         3100.000
CPU max MHz:                     4000.0000
CPU min MHz:                     1200.0000
BogoMIPS:                        6200.00
Virtualization:                  VT-x
L1d cache:                       1.1 MiB
L1i cache:                       1.1 MiB
L2 cache:                        36 MiB
L3 cache:                        49.5 MiB
NUMA node0 CPU(s):               0-17,36-53
NUMA node1 CPU(s):               18-35,54-71
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; TSX disabled
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.9.0            12_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            12_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            12_linux64_mkl    conda-forge
[conda] mkl                       2021.4.0           h06a4308_640    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.4.0            py39h7e14d7c_0    conda-forge
[conda] mkl_fft                   1.3.1            py39h0c7bc48_1    conda-forge
[conda] mkl_random                1.2.2            py39hde0f152_0    conda-forge
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] pytorch                   1.13.1          py3.9_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_3    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.13.1               py39_cu117    pytorch
[conda] torchvision               0.14.1               py39_cu117    pytorch
```

**Comment by angusturner** [comment]
This is not a bug. 

If you wanted both those tensors to be the same, you would need to do:
```
def reset_parameters(self):
        torch.manual_seed(0)
        torch.nn.init.xavier_normal_(self.fc1.weight.data, gain=1)
        torch.manual_seed(0)
        torch.nn.init.xavier_normal_(self.fc2.weight.data, gain=1)
```

**Comment by hannshu** [comment]
Thanks, it really solved my problem! @angusturner
